src/isa.h

```diff
@@ -0,0 +1,256 @@
+/*
+  Copyright (c) 2025, Intel Corporation
+
+  SPDX-License-Identifier: BSD-3-Clause
+*/
+
+#ifndef ISA_H
+#define ISA_H
+
+enum ISA {
+    INVALID = -1,
+
+    SSE2 = 0,
+    SSE41 = 1,
+    SSE42 = 2,
+    AVX = 3,
+    AVX11 = 4,
+    AVX2 = 5,
+    AVX2VNNI = 6,
+    KNL_AVX512 = 7,
+    SKX_AVX512 = 8,
+    ICL_AVX512 = 9,
+    SPR_AVX512 = 10,
+    AVX10_2_512 = 11,
+
+    COUNT
+};
+
+#define UNUSED(x) ((void)(x))
+
+#if defined(__i386__) || defined(__x86_64__) || defined(_M_IX86) || defined(_M_X64)
+
+#if defined(_MSC_VER)
+// __cpuid() and __cpuidex() are defined on Windows in <intrin.h> for x86/x64.
+#include <intrin.h>
+#else
+// On *nix they need to be defined manually through inline assembler.
+static void __cpuid(int info[4], int infoType) {
+    __asm__ __volatile__("cpuid" : "=a"(info[0]), "=b"(info[1]), "=c"(info[2]), "=d"(info[3]) : "0"(infoType));
+}
+
+static void __cpuidex(int info[4], int level, int count) {
+    __asm__ __volatile__("cpuid" : "=a"(info[0]), "=b"(info[1]), "=c"(info[2]), "=d"(info[3]) : "0"(level), "2"(count));
+}
+#endif // !defined(_MSC_VER)
+
+static int xgetbv() {
+#if defined(_MSC_VER)
+    // Check if the OS saves the XMM, YMM and ZMM registers, i.e. it supports AVX2 and AVX512.
+    // See section 2.1 of software.intel.com/sites/default/files/managed/0d/53/319433-022.pdf
+    return _xgetbv(_XCR_XFEATURE_ENABLED_MASK);
+#else
+    // Check xgetbv; this uses a .byte sequence instead of the instruction
+    // directly because older assemblers do not include support for xgetbv and
+    // there is no easy way to conditionally compile based on the assembler used.
+    int rEAX = 0, rEDX = 0;
+    __asm__ __volatile__(".byte 0x0f, 0x01, 0xd0" : "=a"(rEAX), "=d"(rEDX) : "c"(0));
+    return rEAX;
+#endif // !defined(_MSC_VER)
+}
+
+static int __os_has_avx_support() { return (xgetbv() & 6) == 6; }
+
+static int __os_has_avx512_support() {
+#if !defined(MACOS)
```

Comment:
`MACOS` is passed as compiler flag when we compile dispatch.c. But when isa.h is included into ispc.cpp should we use `ISPC_HOST_IS_APPLE`? I'd double check other macros too.

Reply:
> `MACOS` is passed as compiler flag when we compile dispatch.c. But when isa.h is included into ispc.cpp should we use `ISPC_HOST_IS_APPLE`? I'd double check other macros too.

Right, I have decided not to rely on `ISPC_HOST_IS_APPLE` here, but to check `__APPLE__` directly, as they are virtually the same.